Lombiq.Tests.UI/Services/UITestContext.cs

```diff
@@ -103,7 +104,11 @@ internal async Task TriggerAfterPageChangeEventAsync()
             {
                 try
                 {
-                    await afterPageChange.Invoke(this);
+                    // For some reason doing await afterPageChange.Invoke(this) will case exceptions to not propagate
+                    // when there are more than 2 subscribers.
+                    await afterPageChange.GetInvocationList()
+                        .Cast<PageChangeEventHandler>()
+                        .AwaitEachAsync(eventHandler => eventHandler(this));
```

Comment:
Do you know why this was necessary, BTW, @DAud-IcI?

Reply:
As I understand it's not actually safe to call `Invoke` on a multicast delegate and should be avoided. What multicast does, is calling each function in sequence and returning the last one's value. If you think about delegate in a sync context it just has methods that return a certain type so if you call `Invoke` synchronously all it does is start three threads and returns you the last one's `Task` object which the `await` correctly awaits. So all but the last one are essentially fire-and-forget event calls. 

Here is a little experiment to demonstrate the problem:
![image](https://user-images.githubusercontent.com/4007293/152438314-0f06ef93-ca93-4d22-bedd-7fd072656c36.png)

Reply:
I think if we make an InvokeAsync extension method then AsnycFixer should complain about using Invoke. Do you want to do one or both of the following?
1) Expand scope here and make such an exception
2) Open an issue at AsyncFixer that it shouldn't allow Invoke on Task-returning delegates (at least if `+=` is ever used on them).

Reply:
Ah, that makes sense. I actually looked into what multicast delegates do and couldn't figure out the reason - in my understanding, async support is available but it's definite that what actually happens is what you said.

Perhaps we need to do something similar for other events in `UITestExecutionEvents` too. Though the others are actually used in a sync way: https://github.com/Lombiq/UI-Testing-Toolbox/blob/4ebe418a21450c1cf7711cf85bcbeb1eb0cce98e/Lombiq.Tests.UI/Extensions/NavigationWebElementExtensions.cs#L31

Hmm, AsyncFixer didn't complain here, what do you mean? I'd perhaps only add an `MulticastDelegate` extension to do what we have above.

Reply:
In general, if you have a `SomeType.Method()` method and create a `MethodAsync(this SomeType sometype)` extension method then AsyncFixer will give you `error VSTHRD103: condition synchronously blocks. Await MethodAsync instead` whenever a `someType.Method()` call happens.

So if we'd create an InvokeAsync extension for Task returning delegates that would effectively deprecate the regular Invoke method. I realize this is not trivial, becuase you are using a custom delegate and not just `Task<T1,Task>` so this probably deserves its own issue.

Reply:
Yep. Could you please create one here? https://github.com/Lombiq/Helpful-Libraries/issues

Reply:
https://github.com/Lombiq/Helpful-Libraries/issues/115

Reply:
Thanks!